model/mention_map_test.go

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func requireEqualValuesMaps(t *testing.T, one, two map[string][]string) {
+	if len(one) != len(two) {
+		require.FailNow(t, "Map lengths differ")
+	}
+
+	for oneKey, oneValue := range one {
+		twoValue, ok := two[oneKey]
+		if !ok {
+			require.FailNowf(t, "Key %s is present in %v but not in %v", oneKey, one, two)
+		}
+
+		require.ElementsMatch(t, oneValue, twoValue)
+	}
+}
+func requireEqualMentionMaps(t *testing.T, one, two map[string]string) {
+	if len(one) != len(two) {
+		require.FailNow(t, "Map lengths differ")
+	}
+
+	for oneKey, oneValue := range one {
+		twoValue, ok := two[oneKey]
+		if !ok {
+			require.FailNowf(t, "Key %s is present in %v but not in %v", oneKey, one, two)
+		}
+
+		require.Equal(t, oneValue, twoValue)
+	}
+}
+
+func TestUserMentionMapFromURLValues(t *testing.T) {
+	fixture := []struct {
+		values   url.Values
+		expected UserMentionMap
+		error    bool
+	}{
+		{
+			url.Values{},
+			nil,
+			true,
+		},
+		{
+			url.Values{
+				UserMentionsKey:    []string{},
+				UserMentionsIdsKey: []string{},
+			},
+			UserMentionMap{},
+			false,
+		},
+		{
+			url.Values{
+				UserMentionsKey:    []string{"one", "two", "three"},
+				UserMentionsIdsKey: []string{"oneId", "twoId", "threeId"},
+			},
+			UserMentionMap{
+				"one":   "oneId",
+				"two":   "twoId",
+				"three": "threeId",
+			},
+			false,
+		},
+		{
+			url.Values{
+				"wrongKey":         []string{"one", "two", "three"},
+				UserMentionsIdsKey: []string{"oneId", "twoId", "threeId"},
+			},
+			nil,
+			true,
+		},
+		{
+			url.Values{
+				UserMentionsKey: []string{"one", "two", "three"},
+				"wrongKey":      []string{"oneId", "twoId", "threeId"},
+			},
+			nil,
+			true,
+		},
+		{
+			url.Values{
+				UserMentionsKey:    []string{"one", "two"},
+				UserMentionsIdsKey: []string{"justone"},
+			},
+			nil,
+			true,
+		},
+	}
+
+	for _, data := range fixture {
+		actualMap, actualError := UserMentionMapFromURLValues(data.values)
+		if data.error {
+			require.Error(t, actualError)
+			require.Nil(t, actualMap)
+		} else {
+			require.NoError(t, actualError)
+			requireEqualMentionMaps(t, actualMap, data.expected)
```

Comment:
Would `assert.Equal` (or `assert.EqualValues`) by chance already do the "right thing" for the corresponding data structures?

Reply:
It seems that `EqualValues` is only for lists, I was not able to use it with maps. I did not test `Equal`, though, but it would take into account order, which is not safe in maps. If you know a cleaner thing to do this, I am all ears!
If not, I was actually wondering whether move this to `testutils` because it seems kind of useful (I need to use it when testing `mentionsToTeamMembers`, for example.

Reply:
I cannot move that to `testutils` because it creates an import cycle. But I'll go with [`reflect.DeepEqual`](https://golang.org/pkg/reflect/#DeepEqual), which is designed for any type of maps, too.

Reply:
Nevermind, `require.Equal` does the trick. I was completely wrong when I tested this for the first time!

Reply:
Done in https://github.com/mattermost/mattermost-server/pull/13762/commits/552c67e6d459e98b646fd149c792a2bb0eb7486d. The tests for the url.Values needed a bit more of work, as the pairs of {mention, id} may be in different order. The easiest way to test this was to convert the `url.Values` to a `map[string]string`, and while I was coding it, I realized I was repeating the code from the `FromURLValues` function.
So I ended up calling those functions; in order to check that the test is well-formed, I wrote the expected values as `url.Values` too and converted it back to a `map[string]string`, as I did with the actual values.